### PR TITLE
Add support for .diff and .patch

### PR DESCRIPTION
**Pull request by barracks510**
Add the ability to get text-diff and format-patch by adding .diff or
.patch in the end of a commit url. Issue #2641

Obsoletes PR #2846

Signed-off-by: Dennis Chen <barracks510@gmail.com>

### CODE REVIEW

File: models/git_diff.go [hunk 1]
```diff
@@ -418,6 +418,44 @@ func GetDiffRange(repoPath, beforeCommitID string, afterCommitID string, maxLine
 	return diff, nil
 }
 
+func GetRawDiff(repoPath, commitId string, diffType string) (string, error) {
```
Comment (Unknwon) [inline models/git_diff.go]: Can be simplified to `func GetRawDiff(repoPath, commitID, diffType string) (string, error) {`.

File: models/git_diff.go [hunk 2]
```diff
@@ -418,6 +418,44 @@ func GetDiffRange(repoPath, beforeCommitID string, afterCommitID string, maxLine
...
+		if commit.ParentCount() == 0 {
+			cmd = exec.Command("git", "format-patch", "--no-signature", "--stdout", "--root", commitId)
+		} else {
+			c, _ := commit.Parent(0)
+			query := fmt.Sprintf("%s...%s", commitId, c.ID.String())
+			cmd = exec.Command("git", "format-patch", "--no-signature", "--stdout", query)
+		}
+	}
+	cmd.Dir = repoPath
+	cmd.Stderr = os.Stderr
```
Comment (Unknwon) [inline models/git_diff.go]: Use `bytes.Buffer` to collect `os.Stderr`.

File: models/git_diff.go [hunk 3]
```diff
@@ -418,6 +418,44 @@ func GetDiffRange(repoPath, beforeCommitID string, afterCommitID string, maxLine
...
+		return "", err
+	}
+
+	commit, err := repo.GetCommit(commitId)
+	if err != nil {
+		return "", err
+	}
+
+	var cmd *exec.Cmd
+	if diffType == "diff" {
```
Comment (Unknwon) [inline models/git_diff.go]: We should use `select diffType {}` and return error is we do not recognize that type.

File: routers/repo/commit.go [hunk 4]
```diff
@@ -195,7 +195,18 @@ func Diff(ctx *context.Context) {
 }
 
 func RawDiff(ctx *context.Context) {
-	panic("not implemented")
+	userName := ctx.Repo.Owner.Name
```
Comment (Unknwon) [inline routers/repo/commit.go]: Those variables are only been used once, I think we don't need to create them at all, just pass values to the `GetRawDiff` directly.

File: models/git_diff.go (hunk 2 above)
Comment (barracks510) [inline models/git_diff.go]: I believe the precedent is set in Line#395 of `git_diff.go`. Do you have a line of source I could reference to fix this?

Comment (Unknwon) [inline models/git_diff.go]: https://github.com/gogits/git-module/blob/master/command.go#L92

**Comment by barracks510** [comment]
Ah. Sorry. My bad. It must be getting late..

File: models/git_diff.go (hunk 2 above)
Comment (barracks510) [inline models/git_diff.go]: Ok. Awesome. Thanks.

File: models/git_diff.go [hunk 5]
```diff
@@ -418,6 +418,50 @@ func GetDiffRange(repoPath, beforeCommitID string, afterCommitID string, maxLine
...
+	}
+
+	stderr := new(bytes.Buffer)
+
+	cmd.Dir = repoPath
+	cmd.Stderr = stderr
+
+	stdout, err := cmd.Output()
+	if err != nil {
+		return "", fmt.Errorf("Stdout: %v; Stderr: %s", err, stderr.String())
```
Comment (Unknwon) [inline models/git_diff.go]: `err` is not `Stdout`, you can use `fmt.Errorf("%v - %s", err, stderr)`, `%s` calls `String` method automatically.

**Comment by Unknwon** [comment]
I think we need @odinuge 's permission to hijack his PR.

**Comment by odinuge** [comment]
Hi,
Removing my name from my commit is kinda strange.. I can update my PR with a new commit, squashed from all changes in this PR. (But i dont have time for the next 15 hours).

**Comment by Unknwon** [comment]
I found a better way.

1. Save a copy of changed files.
2. Reset to `develop` and fetch commits made by @odinuge 
3. Copy back changed files and commit as @barracks510 

So no one loses credits. Sounds good?

**Comment by odinuge** [comment]
Sounds good @Unknwon!

**Comment by Unknwon** [comment]
@barracks510 if you also agree, please update this PR in proposed way^

**Comment by barracks510** [comment]
Yes. That sounds good.

**Comment by barracks510** [comment]
Sorry, I think I misunderstood. I still need @odinuge to update his patch to the latest develop branch to proceed with the proposed actions, otherwise, it still shows my current commit message.

**Comment by Unknwon** [comment]
That's sad, ping @odinuge

**Comment by odinuge** [comment]
Rebased my commit. See the topic-branch.

**Comment by barracks510** [comment]
Ok. New PR based on Changes. See #3345.

**Pull request closed without merging**